detoks/src/main/java/nl/tudelft/trustchain/detoks/newcoin/AdminWallet.kt

```diff
@@ -4,40 +4,48 @@ import androidx.annotation.RequiresApi
 import nl.tudelft.trustchain.detoks.Token
 import nl.tudelft.trustchain.detoks.db.DbHelper
 
-data class AdminWallet(val tokens: ArrayList<Token>) {
+data class AdminWallet(
+    val publicKey: String
+) {
 
     companion object {
         private var wallet :AdminWallet? = null
-        private val dbHelper: DbHelper? = null
+        private var tokens : MutableList<Token>? = null
+        private var dbHelper: DbHelper? = null
         @RequiresApi(Build.VERSION_CODES.O)
         private fun create(context: Context): AdminWallet {
 //            val generator = KeyPairGenerator.getInstance("RSA")
 //            generator.initialize(2048)
 //            val keyPair = generator.generateKeyPair()
 
-            val dbHelper = DbHelper(context)
-            val tokens = dbHelper.getAllTokens()
+            dbHelper = DbHelper(context)
+            tokens = dbHelper!!.getAllTokens(admin = true)
+            wallet = AdminWallet("s")
 
-            return AdminWallet(tokens)
+            return wallet!!
         }
         @RequiresApi(Build.VERSION_CODES.O)
         fun getInstance(context: Context): AdminWallet {
             return this.wallet ?: create(context)
         }
     }
 
+    fun getTokens(): MutableList<Token> {
+        return tokens!!
+    }
+
     val balance: Int get() {
-        return tokens.size
+        return tokens!!.size
     }
 
     fun addToken(token : Token) {
-        tokens.add(token)
-        dbHelper?.addToken(token)
+        tokens!!.add(token)
+        dbHelper?.addToken(token, admin = true)
     }
```

Comment:
Do check whether the result was actually successfully written in the db with 
if(result != -1L) {
            tokens!!.add(token)
      } 
If it wasn't written successfully in the db, it should not be added to the tokens as well. You can look at the wallet class :)